### ISSUE TITLE
Logic for Separate Log drive?

### ISSUE THREAD

**Issue by DMikulasr** [way0utwest/PowershellScripts#1]
Have you considered including logic to account for log files not being on the same drive as the MDF's?

**Comment by way0utwest** [comment]
No. This is for a specific purpose on simple laptops. Never have these separated.

Sent from my iPhone

On Mar 11, 2015, at 12:49 PM, DMikulasr <notifications@github.com<mailto:notifications@github.com>> wrote:

Have you considered including logic to account for log files not being on the same drive as the MDF's?

##

**Comment by DMikulasr** [comment]
I was playing around with it yesterday and came up with this. I added a
parameter for the L drive first but then remembered as long as the logs are
in the same place the MDF thinks they are you do not need to explicitly
tell SQL where they are. I also add a addition to the $dbname variable to
exclude parts of the MDF that are not supposed to be in the name.

Thank you so much for this script btw, you saved me hours trying to come up
with this on my own!

<#
Description: Attach databases from a folder that are not already on the
instance.
#>

$folder = 'D:\MSSQL\Data'
$Lfolder = 'L:\MSSQL\Data'
$debug = 0
$instance = 'SERVERNAME'
$dbsexist = ""
$dbsnotexist = ""

#placeholder
"================================================="
"=       Start                                   ="
"================================================="

| Out-Null

| Out-Null

| Out-Null

| Out-Null

$server = New-Object ("Microsoft.SqlServer.Management.Smo.Server") $instance
if ($debug -eq 2)
 {
    "Database List"
    "-------------"
    foreach($sqlDatabase in $Server.databases)
        { Write-Output "Debug 2: DB:" $sqlDatabase.name
        }
 #end debug
 }

Write-Output "Checking Files in " + $folder
Write-Output " "

# loop through each  of the file

foreach ($file in Get-ChildItem $folder)
{
#Debug
if ($debug -eq 1)
    {
     Write-Output "Debug 1: All Files: " $file.name
     #end debug
    }

```
# reset the check
$found = 0
```

   if ($file.Extension -eq '.mdf')
    {

```
 if ($debug -eq 1)
 {
  Write-Output "Debug 1: MDF File: " $file.name
  #end debug
 }

 # output file name being checked
 #"File:" + $file.BaseName
 # loop through each  of the databases
 foreach($sqlDatabase in $Server.databases)
    {
     # grab an owner here.
     $dbholder = $server.Databases[0]
     $owner = $dbholder.Owner

     #search the filegroups
     $sqlfg = $sqlDatabase.FileGroups
     foreach ($fg in $sqlfg)
     {
      foreach ($dbfile in $fg.files | Where-Object {$_.ISPrimaryFile
```

-eq $true} )
           {
            #get the mdf file name
            $filebeingchecked = $dbfile.filename

```
       }
     }



     # check the file v the database name
     if ($debug -eq 1)
      {
         "Debug 1: Checking " + $file.FullName + " v " +
```

$filebeingchecked
           }

```
     if ($file.FullName -eq $filebeingchecked)
      {
       $found = 1
       $dbsexist = $dbsexist + $file.BaseName + ", "
       #end if
      }

    }
if ($found -eq 0)
{
    $dbsnotexist = $dbsnotexist + $file.BaseName + ", "
    Write-Output "Need to attach database " + $file.FullName

    # attach the databases
    $dbfiles = New-Object
```

System.Collections.Specialized.StringCollection

```
    $dbfiles.Add($file.FullName) | Out-Null

    #get database name
    $dbname = $file.BaseName
    $dbname = $dbname -replace "_data",""
    # get log file, assuming same basename as mdf
    #$logfile = $lfolder + "\" + $file.BaseName + "_log.ldf"
    #$dbfiles.Add($logfile) | Out-Null

    Write-output "Attaching as database (" + $dbname + ") from mdf (" +
```

$file.FullName + ") "
        try
        {
            $server.AttachDatabase($dbname, $dbfiles)
        #end try
        }
        catch
        {
        Write-Output $_.exception;
        #end catch
        }
        #$Server.AttachDatabase($dbname,$dbfiles, $owner,

```
#end of not found
}

#end if - check extension
```

   }

 #end loop through files
}

if ($debug -eq 1)
{
 Write-Output "Debug 1: Databases that exist: " + $dbsexist
 Write-Output "Debug 1: Databases that don't exist: " + $dbsnotexist
}

Write-Output "Done"

On Wed, Mar 11, 2015 at 8:10 PM, Steve Jones notifications@github.com
wrote:

> No. This is for a specific purpose on simple laptops. Never have these
> separated.
> 
> Sent from my iPhone
> 
> On Mar 11, 2015, at 12:49 PM, DMikulasr <notifications@github.com<mailto:
> notifications@github.com>> wrote:
> 
> Have you considered including logic to account for log files not being on
> the same drive as the MDF's?
> 
> ## 
> 
> Reply to this email directly or view it on GitHub<
> https://github.com/way0utwest/PowershellScripts/issues/1>.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/way0utwest/PowershellScripts/issues/1#issuecomment-78406360
> .

## 

Dennis Mikula
816-316-9490

**Comment by way0utwest** [comment]
I’ll have to give it a try. Thanks for this and you are welcome. It took me a few hours, but it was something that was handy for me to reset my environment easily.
